DatadogCore/Sources/Core/DatadogCore.swift

```diff
@@ -316,29 +303,62 @@ extension DatadogCore: DatadogCoreProtocol {
     }
 }
 
-internal struct DatadogCoreFeatureScope: FeatureScope {
-    let contextProvider: DatadogContextProvider
-    let storage: FeatureStorage
-    let store: FeatureDataStore
+internal class CoreFeatureScope<Feature>: FeatureScope where Feature: DatadogFeature {
+    private weak var core: DatadogCore?
+    private let store: FeatureDataStore
+
+    init(in core: DatadogCore) {
+        self.core = core
+        self.store = FeatureDataStore(
+            feature: Feature.name,
+            directory: core.directory,
+            queue: core.readWriteQueue,
+            telemetry: core.telemetry
+        )
+    }
 
     func eventWriteContext(bypassConsent: Bool, _ block: @escaping (DatadogContext, Writer) -> Void) {
+        // Capture the core reference so it is available until async block completion. This is ensure
+        // that we write events which were collected right before deallocating the core on the current thread.
+        guard let core = core else {
+            return // core is deinitialized
+        }
+
         // (on user thread) request SDK context
         context { context in
-            // (on context thread) call the block
+            // (on context thread)
+            guard let storage = core.stores[Feature.name]?.storage else {
+                if core.get(feature: Feature.self) == nil { // the core was stopped
+                    DD.logger.warn(
+                        "Failed to obtain Event Write Context for '\(Feature.name)' because this feature is no longer running."
+                    )
+                } else { // core is running, but this is wrong Feature type
+                    DD.logger.error(
+                        "Failed to obtain Event Write Context for '\(Feature.name)' because it is not a `DatadogRemoteFeature`."
+                    )
+                    #if DEBUG
+                    assertionFailure("Obtaining Event Write Context for '\(Feature.name)' but it is not a `DatadogRemoteFeature`.")
+                    #endif
+                }
+                return
+            }
```

Comment:
The `storage` you mean? Yes we could - `core.stores` is already behind `@ReadWriteLock` so it is safe. However, the consequence will be that EWC could not be used within `DatadogFeature.init()`, e.g.:
```swift
class DemoFeature: DatadogRemoteFeature {
   init(in core: DatadogCoreProtocol) {
      // ...

      let scope = core.scope(for: DemoFeature.self)
      scope.eventWriteContext { _, writer in
         writer.write("initial event")
      }
   }
}
```
Reason being the `DemoFeature` being not yet registered, so it has yet no `storage` created in core. In the design proposed in this PR this is not the case because we jump on queue, so the `storage` gets registered before we ask for it. 🤔💭 I wonder tho if this is not exposing race condition.

For the sake of simplicity, and because we have no "initial event" use case anywhere, I'm okay to capture the `storage` as you propose. That will also prevent from extending the `core` lifecycle for EWC 👍.

Reply:
Changed ✅